### ISSUE TITLE
CA1508 mis-fires under belief that second subexpression is unreachable

### ISSUE THREAD

**Issue by AArnott** [dotnet/roslyn-analyzers#5737]
### Analyzer

**Diagnostic ID**: [CA1508](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1508): `Avoid dead conditional code`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 6.0.100](https://dotnet.microsoft.com/download/dotnet/6.0)

### Describe the bug

CA1508 misfires on the second half of the below expression, even though no code preceding this in the method makes a guarantee around the Count property on this collection.

```cs
if (this.issuedUpgradeableReadLocks.Count == 0 && this.issuedWriteLocks.Count == 0)
```

[Full context](https://github.com/microsoft/vs-threading/pull/962/commits/9075367c3a1c0fb5e8e2232cade32b28ddd9f0fd#diff-8112250e1e17da88e008cc6c9b35abc46fed2874053b2219868ca488ab72c531)

### Steps To Reproduce

```
git clone https://github.com/AArnott/vs-threading.git
cd vs-threading
git checkout 9075367c3a1c0fb5e8e2232cade32b28ddd9f0fd~
msbuild /r
```

### Expected behavior

No warnings.

### Actual behavior

Warnings about dead code.

**Comment by Youssef1313** [comment]
Related: https://github.com/dotnet/roslyn-analyzers/issues/5789

**Comment by mavasani** [comment]
I am able to repro this on 7.0.0 version of Microsoft.CodeAnalysis.NetAnalyzers package with the below minimal repro case, but not on the latest 8.x.x version in the .NET 8 preview SDK (`C:\Program Files\dotnet\sdk\8.0.100-preview.6.23330.14\Sdks\Microsoft.NET.Sdk\analyzers\Microsoft.CodeAnalysis.NetAnalyzers.dll`):

```csharp
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace CA1508
{
    class C
    {
        private readonly HashSet<C> issuedWriteLocks = new HashSet<C>();
        private readonly HashSet<C> issuedReadLocks = new HashSet<C>();
        
        void M() { }

        void M2()
        {
            if (this.issuedReadLocks.Count == 0 && this.issuedWriteLocks.Count == 0)
            {
            }
            else
            {
                this.M();
                if (this.issuedReadLocks.Count == 0 && this.issuedWriteLocks.Count == 0) // CA1508 on 'this.issuedWriteLocks.Count == 0'
                {
                }
            }
        }
    }
}
```
